File: app/src/main/java/com/masterlibs/basestructure/utils/ImageToPdf.java
package com.masterlibs.basestructure.utils;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;
import android.os.ParcelFileDescriptor;
import android.util.Log;

import com.documentmaster.documentscan.Const;
import com.documentmaster.documentscan.OnActionCallback;
import com.itextpdf.text.Document;
import com.itextpdf.text.Image;
import com.itextpdf.text.PageSize;
import com.itextpdf.text.Rectangle;
import com.itextpdf.text.pdf.PdfWriter;
import com.masterlibs.basestructure.R;
import com.masterlibs.basestructure.model.ItemFile;
import com.shockwave.pdfium.PdfDocument;
import com.shockwave.pdfium.PdfiumCore;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.util.Arrays;
import java.util.List;

@SuppressLint("StaticFieldLeak")
public class ImageToPdf extends AsyncTask<String, String, String> {
    private static final String ITEM_ALL_FILE = "ITEM_ALL_FILE";
    private final ProgressDialog dialog;
    private final List<String> imagesUri;
    private final Context context;
    private final OnActionCallback callback;
    private String path;
    private String fileName;

    public ImageToPdf(List<String> imagesUri, Context context, OnActionCallback callback) {
        this.callback = callback;
        this.imagesUri = imagesUri;
        this.context = context;
        dialog = new ProgressDialog(context);
        dialog.setCancelable(false);
        dialog.setMessage(context.getString(R.string.converting) + "... 0%");
        dialog.setButton(ProgressDialog.BUTTON_NEGATIVE, context.getString(R.string.cancel), (dialog, which) -> {
            ImageToPdf.this.cancel(true);
            if (path != null)
                new File(path).delete();
            dismissProgressDialog();
        });
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        dialog.show();
    }

    @Override
    protected String doInBackground(String... params) {
        path = Environment.getExternalStorageDirectory().toString() + "/Documents/" + Arrays.toString(params) + ".pdf";
        long totalSize = getTotalSize();
        long currentConvert = 0;
        Document document = new Document(PageSize.A4, 38, 38, 50, 38);
        Rectangle documentRect = document.getPageSize();
        try {
            PdfWriter.getInstance(document, new FileOutputStream(path));
            document.open();
            for (int i = 0; i < imagesUri.size(); i++) {


                Bitmap bmp;
                bmp = BitmapFactory.decodeFile(imagesUri.get(i));
                ByteArrayOutputStream stream = new ByteArrayOutputStream();
                bmp.compress(Bitmap.CompressFormat.PNG, 80, stream);

                currentConvert += new File(imagesUri.get(i)).length();
                int percent = (int) ((currentConvert * 1.0 / totalSize) * 100);
                publishProgress(percent + "%", imagesUri.get(i));


                Image image = Image.getInstance(stream.toByteArray());
                if (bmp.getWidth() > documentRect.getWidth() || bmp.getHeight() > documentRect.getHeight()) {
                    image.scaleToFit(documentRect.getWidth() - 16, documentRect.getHeight() - 16);
                } else {
                    image.scaleToFit(bmp.getWidth(), bmp.getHeight());
                }
                image.setAbsolutePosition((documentRect.getWidth() - image.getScaledWidth()) / 2,
                        (documentRect.getHeight() - image.getScaledHeight()) / 2);
                image.setBorder(Image.BOX);
                document.add(image);

                document.newPage();
            }
            document.close();
            File file = new File(path);
            int totalPage = getNumberPage(file) - 1;
//            try {
//                categoryList.get(ITEM_ALL_FILE).addFile(0, file, totalPage);
//            } catch (Exception e) {
//                e.printStackTrace();
//            }

        } catch (Exception e) {
            Log.d("TAG", "doInBackground: " + e.getMessage());
            e.printStackTrace();
        }
        imagesUri.clear();
        return null;
    }

    private int getNumberPage(File file) {
        PdfiumCore pdfiumCore = new PdfiumCore(context);
        try {
            ParcelFileDescriptor fd = context.getContentResolver().openFileDescriptor(Uri.fromFile(file), "r");
            PdfDocument pdfDocument = pdfiumCore.newDocument(fd);
            return pdfiumCore.getPageCount(pdfDocument);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return 0;
    }

    @Override
    protected void onProgressUpdate(String... values) {
        super.onProgressUpdate(values);
        String percent = values[0];
        //callback.callback(Const.PATH_PDF, path);

        //dialog.setMessage("Creating..." + percent);
    }

    private long getTotalSize() {
        long rs = 0;
        for (int i = 0; i < imagesUri.size(); i++) {
            rs += new File(imagesUri.get(i)).length();
        }
        return rs;
    }

    @Override
    protected void onPostExecute(String s) {
        super.onPostExecute(s);
        if (((Activity) context).isDestroyed())
            return;
        dismissProgressDialog();
        try {
            ItemFile itemFile = new ItemFile(path);
            callback.callback(Const.PATH_PDF, path);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    void dismissProgressDialog() {
        if (dialog != null && dialog.isShowing())
            dialog.dismiss();
    }
}

File: app/src/main/java/com/masterlibs/basestructure/task/ImageToPdf.kt
package com.masterlibs.basestructure.task

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.AsyncTask
import android.util.Log
import com.documentmaster.documentscan.OnActionCallback
import com.itextpdf.text.Document
import com.itextpdf.text.Image
import com.itextpdf.text.PageSize
import com.itextpdf.text.pdf.PdfWriter
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream

@SuppressLint("StaticFieldLeak")
class ImageToPdf(
    private val imagesUri: MutableList<String>,
    private val context: Context,
    private val callback: OnActionCallback
) : AsyncTask<String?, String?, String?>() {
    private val dialog: ProgressDialog?
    private var path: String? = null
    override fun onPreExecute() {
        super.onPreExecute()
        dialog!!.show()
    }

    override fun doInBackground(vararg params: String?): String? {
        path = params[0]
        val totalSize = totalSize
        var currentConvert: Long = 0
        val document = Document(PageSize.A4, 38F, 38F, 50F, 38F)
        val documentRect = document.pageSize
        try {
            PdfWriter.getInstance(document, FileOutputStream(path))
            document.open()
            for (i in imagesUri.indices) {
                var bmp: Bitmap
                bmp = BitmapFactory.decodeFile(imagesUri[i])
                val stream = ByteArrayOutputStream()
                bmp.compress(Bitmap.CompressFormat.PNG, 80, stream)
                currentConvert += File(imagesUri[i]).length()
                val percent = (currentConvert * 1.0 / totalSize * 100).toInt()
                publishProgress("$percent%", imagesUri[i])
                val image = Image.getInstance(stream.toByteArray())
                if (bmp.width > documentRect.width || bmp.height > documentRect.height) {
                    image.scaleToFit(documentRect.width - 16, documentRect.height - 16)
                } else {
                    image.scaleToFit(bmp.width.toFloat(), bmp.height.toFloat())
                }
                image.setAbsolutePosition(
                    (documentRect.width - image.scaledWidth) / 2,
                    (documentRect.height - image.scaledHeight) / 2
                )
                image.border = Image.BOX
                document.add(image)
                document.newPage()
            }
            document.close()
            val file = File(path)
        } catch (e: Exception) {
            Log.d("TAG", "doInBackground: " + e.message)
            e.printStackTrace()
        }
        imagesUri.clear()
        return null
    }

    override fun onProgressUpdate(vararg values: String?) {
        super.onProgressUpdate(*values)
        val percent = values[0]
        dialog!!.setMessage("Creating...$percent")
    }

    private val totalSize: Long
        private get() {
            var rs: Long = 0
            for (i in imagesUri.indices) {
                rs += File(imagesUri[i]).length()
            }
            return rs
        }

    override fun onPostExecute(s: String?) {
        super.onPostExecute(s)
        if ((context as Activity).isDestroyed) return
        dismissProgressDialog()
        try {
            callback.callback(null, path)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun dismissProgressDialog() {
        if (dialog != null && dialog.isShowing) dialog.dismiss()
    }

    init {
        dialog = ProgressDialog(context)
        dialog.setCancelable(false)
        dialog.setMessage("Creating... 0%")
    }
}